// zipkin/lib/annotation.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var InetAddress = require('./InetAddress');

var SimpleAnnotation = function () {
  function SimpleAnnotation() {
    _classCallCheck(this, SimpleAnnotation);
  }

  _createClass(SimpleAnnotation, [{
    key: 'toString',
    value: function toString() {
      return this.annotationType + '()';
    }
  }]);

  return SimpleAnnotation;
}();

var ClientSend = function (_SimpleAnnotation) {
  _inherits(ClientSend, _SimpleAnnotation);

  function ClientSend() {
    _classCallCheck(this, ClientSend);

    return _possibleConstructorReturn(this, (ClientSend.__proto__ || Object.getPrototypeOf(ClientSend)).apply(this, arguments));
  }

  return ClientSend;
}(SimpleAnnotation);

var ClientRecv = function (_SimpleAnnotation2) {
  _inherits(ClientRecv, _SimpleAnnotation2);

  function ClientRecv() {
    _classCallCheck(this, ClientRecv);

    return _possibleConstructorReturn(this, (ClientRecv.__proto__ || Object.getPrototypeOf(ClientRecv)).apply(this, arguments));
  }

  return ClientRecv;
}(SimpleAnnotation);

var ServerSend = function (_SimpleAnnotation3) {
  _inherits(ServerSend, _SimpleAnnotation3);

  function ServerSend() {
    _classCallCheck(this, ServerSend);

    return _possibleConstructorReturn(this, (ServerSend.__proto__ || Object.getPrototypeOf(ServerSend)).apply(this, arguments));
  }

  return ServerSend;
}(SimpleAnnotation);

var ServerRecv = function (_SimpleAnnotation4) {
  _inherits(ServerRecv, _SimpleAnnotation4);

  function ServerRecv() {
    _classCallCheck(this, ServerRecv);

    return _possibleConstructorReturn(this, (ServerRecv.__proto__ || Object.getPrototypeOf(ServerRecv)).apply(this, arguments));
  }

  return ServerRecv;
}(SimpleAnnotation);

var ProducerStart = function (_SimpleAnnotation5) {
  _inherits(ProducerStart, _SimpleAnnotation5);

  function ProducerStart() {
    _classCallCheck(this, ProducerStart);

    return _possibleConstructorReturn(this, (ProducerStart.__proto__ || Object.getPrototypeOf(ProducerStart)).apply(this, arguments));
  }

  return ProducerStart;
}(SimpleAnnotation);

var ProducerStop = function (_SimpleAnnotation6) {
  _inherits(ProducerStop, _SimpleAnnotation6);

  function ProducerStop() {
    _classCallCheck(this, ProducerStop);

    return _possibleConstructorReturn(this, (ProducerStop.__proto__ || Object.getPrototypeOf(ProducerStop)).apply(this, arguments));
  }

  return ProducerStop;
}(SimpleAnnotation);

var ConsumerStart = function (_SimpleAnnotation7) {
  _inherits(ConsumerStart, _SimpleAnnotation7);

  function ConsumerStart() {
    _classCallCheck(this, ConsumerStart);

    return _possibleConstructorReturn(this, (ConsumerStart.__proto__ || Object.getPrototypeOf(ConsumerStart)).apply(this, arguments));
  }

  return ConsumerStart;
}(SimpleAnnotation);

var ConsumerStop = function (_SimpleAnnotation8) {
  _inherits(ConsumerStop, _SimpleAnnotation8);

  function ConsumerStop() {
    _classCallCheck(this, ConsumerStop);

    return _possibleConstructorReturn(this, (ConsumerStop.__proto__ || Object.getPrototypeOf(ConsumerStop)).apply(this, arguments));
  }

  return ConsumerStop;
}(SimpleAnnotation);

function LocalOperationStart(name) {
  this.name = name;
}
LocalOperationStart.prototype.toString = function () {
  return 'LocalOperationStart("' + this.name + '")';
};

var LocalOperationStop = function (_SimpleAnnotation9) {
  _inherits(LocalOperationStop, _SimpleAnnotation9);

  function LocalOperationStop() {
    _classCallCheck(this, LocalOperationStop);

    return _possibleConstructorReturn(this, (LocalOperationStop.__proto__ || Object.getPrototypeOf(LocalOperationStop)).apply(this, arguments));
  }

  return LocalOperationStop;
}(SimpleAnnotation);

function Message(message) {
  this.message = message;
}
Message.prototype.toString = function () {
  return 'Message("' + this.message + '")';
};

function ServiceName(serviceName) {
  this.serviceName = serviceName;
}
ServiceName.prototype.toString = function () {
  return 'ServiceName("' + this.serviceName + '")';
};

function Rpc(name) {
  this.name = name;
}
Rpc.prototype.toString = function () {
  return 'Rpc("' + this.name + '")';
};

function ClientAddr(_ref) {
  var host = _ref.host,
      port = _ref.port;

  this.host = host;
  this.port = port;
}
ClientAddr.prototype.toString = function () {
  return 'ClientAddr(host="' + this.host + '", port=' + this.port + ')';
};

function ServerAddr(_ref2) {
  var serviceName = _ref2.serviceName,
      host = _ref2.host,
      port = _ref2.port;

  this.serviceName = serviceName;
  this.host = host || undefined;
  this.port = port || 0;
}
ServerAddr.prototype.toString = function () {
  return 'ServerAddr(serviceName="' + this.serviceName + '", host="' + this.host + '", port=' + this.port + ')';
};

function LocalAddr(_ref3) {
  var host = _ref3.host,
      port = _ref3.port;

  this.host = host || InetAddress.getLocalAddress();
  this.port = port || 0;
}
LocalAddr.prototype.toString = function () {
  return 'LocalAddr(host="' + this.host.toString() + '", port=' + this.port + ')';
};

function MessageAddr(_ref4) {
  var serviceName = _ref4.serviceName,
      host = _ref4.host,
      port = _ref4.port;

  this.serviceName = serviceName;
  this.host = host;
  this.port = port;
}
MessageAddr.prototype.toString = function () {
  return 'MessageAddr(serviceName="' + this.serviceName + '", host="' + this.host + '", port=' + this.port + ')';
};

function BinaryAnnotation(key, value) {
  this.key = key;
  this.value = value;
}
BinaryAnnotation.prototype.toString = function () {
  return 'BinaryAnnotation(' + this.key + '="' + this.value + '")';
};

var annotation = {
  ClientSend: ClientSend,
  ClientRecv: ClientRecv,
  ServerSend: ServerSend,
  ServerRecv: ServerRecv,
  ProducerStart: ProducerStart,
  ProducerStop: ProducerStop,
  ConsumerStart: ConsumerStart,
  ConsumerStop: ConsumerStop,
  MessageAddr: MessageAddr,
  Message: Message,
  ServiceName: ServiceName,
  Rpc: Rpc,
  ClientAddr: ClientAddr,
  ServerAddr: ServerAddr,
  LocalAddr: LocalAddr,
  BinaryAnnotation: BinaryAnnotation,
  LocalOperationStart: LocalOperationStart,
  LocalOperationStop: LocalOperationStop
};

Object.keys(annotation).forEach(function (key) {
  annotation[key].prototype.annotationType = key;
});

module.exports = annotation;
